### ISSUE TITLE
implement blocktree data structure

### ISSUE THREAD

**Issue by noot** [ChainSafe/mina-rs#30]
**Task summary**

 implement blocktree data structure which tracks the different forks of the network in a tree structure starting from a root node (genesis) and branching out. the nodes should be of type `ProtocolState` and each node should have only 1 parent (its parent block), but each node can have any number of children

**Specification reference**

- this will be useful for chain selection, as it will help us easily track all the forks in the network in one structure: https://github.com/jspada/mina/tree/8e6558e5f698cb01f5f99a45dc097d04729f7128/docs/specs/consensus#42-select-chain

**Other information and links**

 found this package, might be handy: https://docs.rs/crate/trees/0.4.2 or feel free to use any other tree package that implements what we need

**Comment by jspada** [comment]
Rather than root the tree from genesis, don't we want to root it to the most recent succinct state?

**Comment by noot** [comment]
> Rather than root the tree from genesis, don't we want to root it to the most recent succinct state?

@jspada yes exactly, I just meant when we initialize the node it starts at genesis. when it syncs, the root of the blocktree will be moved along to the succint root that's 290 blocks behind the highest block like you mentioned. I don't think we'd need to unprune, as past 290 blocks the state should be considered probabilistically finalized, but we can double check.
